fairscale/optim/oss.py

```diff
@@ -201,6 +202,72 @@ def step(self, closure: Optional[Callable[[], float]] = None, **kwargs: Any) ->
 
         return loss
 
+    def clip_grad_norm(self, max_norm: float, norm_type: float = 2.0) -> torch.Tensor:
+        """
+        Clip all gradients at this point in time. The norm is computed over all gradients together, as if they were
+        concatenated into a single vector. Gradients are modified in-place.
+
+        Arguments:
+            max_norm (float or int): max norm of the gradients
+            norm_type (float or int): type of the used p-norm. Can be ``'inf'`` for
+                infinity norm.
+
+        Returns:
+            Total norm of the parameters (viewed as a single vector).
+
+        .. note: This is analogous to `torch.nn.utils.clip_grad_norm_` but handles the partitioning and multiple devices per rank
+        under the hood.
+
+        .. warning: This needs to be called on all ranks, since synchronization primitives will be used
+        """
+
+        # Compute the max norm for this shards's worth of gradients
+        max_norm = float(max_norm)
+        norm_type = float(norm_type)
+
+        # Filter out the grad-less params, concatenate params from all devices
+        local_params = [
+            filter(lambda x: x.grad is not None, device_params[self.rank])
+            for device_params in self.per_device_params.values()
+        ]
+
+        # Compute the norm on this grad set
+        if norm_type == inf:
+            local_norm = max(
+                p.grad.detach().abs().max().to(self._device)  # type: ignore   # mypy seems wrong here, None grads have been filtered out
+                for local_params_device in local_params
+                for p in local_params_device
+            )
+
+        else:
+            local_norm = torch.norm(  # type: ignore
+                torch.stack(
+                    [
+                        torch.norm(p.grad.detach(), norm_type).to(self._device)  # type: ignore
+                        for local_params_device in local_params
+                        for p in local_params_device
+                    ]
+                ),
+                norm_type,
+            )
+
+        # Sync all the norms from all ranks
+        norms = [torch.zeros_like(local_norm) for k in range(self.world_size)]
+        dist.all_gather(norms, local_norm, group=self.group, async_op=False)
+        total_norm = torch.max(torch.stack(norms))
```

Comment:
I don't think the logic above is correct. For regular norms, you could do something like this:
```suggestion
        total_norm = local_norm ** norm_type
        total_norm = dist.all_reduce(total_norm, group=self.group)
        total_norm = total_norm ** (1/norm_type)
```
You would need different logic for inf norm.

Reply:
@msbaines: Is this correct? `self.group` is the Data Parallel group -- we shouldn't be summing across data parallel workers for the clip norm. Instead, we should be summing across model parallel units. 

When making this change, it's important not to overcount. One way is to flag each model parallel weight with a bit saying whether it's part of a model parallel calculation or just a repeated parameter.

Consider these two trivial examples:
1) Model parallel units don't have any model parallel operations; they have the exact same parameters. When computing the clip norm, if you allreduce over model parallel units, you'll double count!
2) Model parallel units only have model parallel operations. In this setting, you're okay.

Reply:
This feedback was specific to Data Parallel. Normally, after DDP backward, each worker has access to all_reduced gradients for every parameter. So a local total_norm calculation will be consistent across workers. With optimizer state sharding, you can avoid all_reduce-ing gradients and instead just reduce on the worker with the shard. But if you do that, workers no longer have access to all gradients. That is the rationale for an OSS-specific clip_grads.

To get a correct total_norm, each worker needs to calculate it own norm and then combine it with the local_norms from other workers to get a total_norm.

Reply:
@msbaines you're right, I had inf in mind but that's mostly wrong, fixing this